python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -0,0 +1,780 @@
+        self._wss_key = wss_key
+
+        if self._wss_key not in self.WSS_KEY_CACHE:
+            self.WSS_KEY_CACHE[self._wss_key] = dict()
+
+        # Cache path on disk.
+        self._cache_path = os.path.join(
+            self._engine.cache_location, 
+            "shotgun_engine_commands_v%s.sqlite" % self.DATABASE_FORMAT_VERSION,
+        )
+
+    ###########################################################################
+    # Properties
+
+    @property
+    def logger(self):
+        """
+        The associated engine's logger.
+        """
+        return self._engine.logger
+
+    @property
+    def host(self):
+        """
+        The host associated with the currnt RPC transaction.
+        """
+        return self._host
+
+    ###########################################################################
+    # Public methods
+
+    def execute_action(self, data):
+        """
+        Executes the engine command associated with the action triggered
+        in the client.
+
+        :param dict data: The payload from the client.
+        """
+        if data["name"] == "__clone_pc":
+            self.logger.debug("Clone configuration command received.")
+            try:
+                self._clone_configuration(data)
+            except Exception as e:
+                self.host.reply(
+                    dict(
+                        retcode=self.COMMAND_FAILED,
+                        err=str(e),
+                        out=str(e),
+                    ),
+                )
+                raise
+            else:
+                self.logger.debug("Clone configuration successful.")
+                self.host.reply(
+                    dict(
+                        retcode=self.COMMAND_SUCCEEDED,
+                        err="",
+                        out="",
+                    ),
+                )
+                return
+
+        project_entity, entities = self._get_entities_from_payload(data)
+
+        # It shouldn't ever happen at this stage of the workflow, but we'll
+        # check to make sure that we can get an environment for the type of
+        # we're dealing with here.
+        env_name = self.__pick_environment(project_entity, entities[0])
+
+        if env_name is None:
+            self.logger.error(
+                "The provided entity does not correspond to any environment."
+            )
+            return
+
+        args_file = self._get_arguments_file(
+            dict(
+                config=data["pc"],
+                name=data["name"],
+                entities=entities,
+                project=project_entity,
+                sys_path=sys.path,
+                base_configuration=self.BASE_CONFIG_URI,
+                engine_name=self.ENGINE_NAME,
+            ),
+        )
+
+        script = os.path.join(
+            os.path.dirname(__file__),
+            "scripts",
+            "execute_command.py"
+        )
+
+        output = None
+
+        try:
```

Comment:
logging timings here would be great for support and debugging. can you use http://developer.shotgunsoftware.com/tk-core/utils.html#sgtk.log.LogManager.log_timing ?